### PR TITLE
improvent handling of changed customizer preset

### PR DESCRIPTION
**Pull request by MichaelPFrey**
*  if current preset is changed, ask user if he wants to save the current preset 
* indicate the the current preset is changed with an asterisks
see also #2316
* move common code to parametervirtualwidget.cpp
* set the parent for parametewidgets
* showDescription/DescriptionLoD (Level of Detail)
   * unify the naming
   * turn into enum

Todo:
- [x] merge #2330 with master, merge master with this
- [x] reduce scope of the variables
- [x] Testing
- [x] when precisely has _valueChanged_ to be changed? -> covered by testing
- [x] write a short warning, that comboBoxPreset->setCurrentIndex(lastComboboxIndex); can cause a recursion
-  ~~disable the save button when there is nothing to save?~~ -> NO. User might want to save when for e.g. parameters where added in the SCAD file, that are not yet present in the json file

### CODE REVIEW

File: src/parameter/parametervirtualwidget.h [hunk 1]
```diff
@@ -5,6 +5,8 @@
 #include "ui_ParameterEntryWidget.h"
 #include "parameterobject.h"
 
+enum descLoD {ShowDetails,Inline,HideDetails,DescOnly};
```
Comment (kintel) [inline src/parameter/parametervirtualwidget.h]: Prefer `enum class`, to get some extra type safety.
..and Capitalize the first letter for good measure.

File: src/parameter/parametervirtualwidget.h [hunk 2]
```diff
@@ -13,7 +15,7 @@ class ParameterVirtualWidget : public QWidget, public Ui::ParameterEntryWidget
 	ParameterObject *object;
 
 public:
-	ParameterVirtualWidget(QWidget *parent = nullptr);
+	ParameterVirtualWidget(QWidget *parent,ParameterObject *parameterobject, int descriptionLoD);
```
Comment (kintel) [inline src/parameter/parametervirtualwidget.h]: This `int` should really be the enum you just defined.

File: src/parameter/parametervirtualwidget.cpp [hunk 3]
```diff
@@ -11,6 +12,24 @@ ParameterVirtualWidget::ParameterVirtualWidget(QWidget *parent) : QWidget(parent
 	policy.setHorizontalStretch(0);
 	policy.setVerticalStretch(0);
 	this->setSizePolicy(policy);
+
+	this->object = parameterobject;
```
Comment (kintel) [inline src/parameter/parametervirtualwidget.cpp]: As this is a pure assignment, we can clarify this by instead adding it to the initialization list for this constructor.

File: src/parameter/parameterset.h [hunk 4]
```diff
@@ -25,6 +25,6 @@ class ParameterSet
 	bool readParameterSet(const std::string &filename);
 	void writeParameterSet(const std::string &filename);
 	void applyParameterSet(class FileModule *fileModule, const std::string &setName);
-	const bool isEmpty(){return root.empty();}
+	bool isEmpty() const{return root.empty();};
```
Comment (kintel) [inline src/parameter/parameterset.h]: stray semicolon.
Some spacing wouldn't hurt either; `bool isEmpty() const { return root.empty(); }`

File: src/parameter/ParameterWidget.cc [hunk 5]
```diff
@@ -188,8 +207,33 @@ void ParameterWidget::setComboBoxPresetForSet()
 	}
 }
 
+//set selection
 void ParameterWidget::onSetChanged(int idx)
 {
+	static int lastComboboxIndex(0);
```
Comment (kintel) [inline src/parameter/ParameterWidget.cc]: Not sure I understand the use of `static` here.
e.g. what happens if you have two OpenSCAD windows open with two different files and two customizers?

File: src/parameter/parametervirtualwidget.h (hunk 1 above)
Comment (MichaelPFrey) [reply src/parameter/parametervirtualwidget.h]: good idea. It results in just one awkward line, where I have to static_cast comboBoxDetails->currentIndex() - but that´s fine as static_cast is precisely the right indicator for whats going on.

File: src/parameter/ParameterWidget.cc (hunk 5 above)
Comment (MichaelPFrey) [reply src/parameter/ParameterWidget.cc]: opps... it feels wrong to have a member variable that is only supposed to be used by one function, but I guess a member variable is the best solution after all.

File: src/parameter/ParameterWidget.cc [hunk 6]
```diff
@@ -422,8 +478,8 @@ void ParameterWidget::writeParameterSets()
 {
 	if(this->unreadableFileExists){
 		QMessageBox msgBox;
-		msgBox.setText(_("Saving presets"));
-		msgBox.setInformativeText(QString(_("%1 was found, but was unreadble. Do you want to overwrite %1?")).arg(QString::fromStdString(this->jsonFile)));
+		msgBox.setWindowTitle(_("Saving presets"));
+		msgBox.setText(QString(_("%1 was found, but was unreadble. Do you want to overwrite %1?")).arg(QString::fromStdString(this->jsonFile)));
```
Comment (kintel) [inline src/parameter/ParameterWidget.cc]: Typo: s/unreadble/unreadable/

**Pull request merged**